system_metrics_collector/test/topic_statistics_collector/test_subscriber_topic_statistics.cpp

```diff
@@ -0,0 +1,346 @@
+constexpr const int64_t kConstantTimestamp = 1000000;
+constexpr const char kPublishPeriodParam[] = "publish_period";
+constexpr const std::chrono::milliseconds kTestDuration{250};
+constexpr const char kTestNodeName[] = "test_periodic_node";
+constexpr const char kTestMetricName[] = "test_metric_name";
+constexpr const char kTestMetricUnit[] = "test_unit";
+constexpr const char kTestTopic[] = "/test_topic";
+}  // namespace
+
+class TestSubscriberTopicStatistics
+  : public topic_statistics_collector::SubscriberTopicStatistics<
+    sensor_msgs::msg::Imu>
+{
+public:
+  TestSubscriberTopicStatistics(const std::string & name, const rclcpp::NodeOptions & options)
+  : SubscriberTopicStatistics<sensor_msgs::msg::Imu>{name, options, kTestTopic} {}
+
+  ~TestSubscriberTopicStatistics() = default;
+
+  std::string GetMetricName() const override
+  {
+    return kTestMetricName;
+  }
+
+  const std::string & GetMetricUnit() const override
+  {
+    static const std::string unit_name{kTestMetricUnit};
+    return unit_name;
+  }
+
+  /**
+   * Return true if the lifecycle publisher is activated, false if null or not activated.
+   *
+   * @return
+   */
+  bool IsPublisherActivated() const
+  {
+    return publisher_ != nullptr && publisher_->is_activated();
+  }
+
+  int GetNumPublished() const
+  {
+    return times_published_;
+  }
+
+private:
+  /**
+   * Test publish for the fixture.
+   */
+  void PublishStatisticMessage() override
+  {
+    ++times_published_;
+  }
+
+  std::atomic<int> times_published_{0};
+};
+
+/**
+ * Node to publish messages on the test topic to trigger subscriber callback.
+*/
+class ImuMessagePublisher : public rclcpp::Node
+{
+public:
+  ImuMessagePublisher()
+  : Node("imu_publisher"), publisher_(nullptr)
+  {
+    auto qos_options = rclcpp::QoS(rclcpp::KeepAll());
+    auto publisher_options = rclcpp::PublisherOptions();
+
+    publisher_ = create_publisher<sensor_msgs::msg::Imu>(
+      kTestTopic,
+      qos_options,
+      publisher_options);
+  }
+
+  ~ImuMessagePublisher() = default;
+
+  void publish(std::shared_ptr<sensor_msgs::msg::Imu> msg)
+  {
+    publisher_->publish(*msg);
+  }
+
+private:
+  typename rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr publisher_;
+};
+
+/**
+ * Test fixture to test a SubscriberTopicStatistics node
+ */
+class SubscriberTopicStatisticsTestFixture : public testing::Test
+{
+public:
+  void SetUp() override
+  {
+    rclcpp::init(0, nullptr);
+
+    rclcpp::NodeOptions options;
+    options.append_parameter_override(kPublishPeriodParam, kDontPublishDuringTest.count());
+
+    test_periodic_publisher_ = std::make_shared<TestSubscriberTopicStatistics>(
+      kTestNodeName, options);
+
+    const moving_average_statistics::StatisticData data =
+      test_periodic_publisher_->GetStatisticsResults();
+    ASSERT_TRUE(std::isnan(data.average));
+    ASSERT_TRUE(std::isnan(data.min));
+    ASSERT_TRUE(std::isnan(data.max));
+    ASSERT_TRUE(std::isnan(data.standard_deviation));
```

Comment:
Does `std::isfinite` make sense here?

Reply:
The contract in [moving_average](https://github.com/ros-tooling/system_metrics_collector/blob/e8bcb98837427268c325901d44a4a47126385591/system_metrics_collector/src/moving_average_statistics/moving_average.hpp#L43) is to return a `NaN` when no observations are made. Hence the use of `std::isnan()`.